network/gopubsub/pubsub.go

```diff
@@ -0,0 +1,126 @@
+package gopubsub
+
+import (
+	"fmt"
+	"strconv"
+	"sync"
+)
+
+//PubSubChannel represents a pubsub system where subscriber can .Subscribe() and publishers can .Publish() or .Close().
```

Comment:
Great, I will take a look at it.

Reply:
I have read `SubscribeToNeighbourhoodDepthChange` and `SubscribePull`, it is the initial implementation I used, but it has a problem, first, is not extracted to a resuable file, and also, it has the problem that the subscriber can not stop the subscription from their side. 
In PubSubChannel, whenever the subscriber wants to stop the subscription, the channel is marked for closing, but not actually closed (because the only one closing a channel should be the writer of the channel).
I think those two pieces of code should be migrated to PubSubChannel for code sanity.

Reply:
> not extracted to a resuable file

I'm not sure this little amount of code justifies being generalized into a component in its own

> the subscriber can not stop the subscription from their side

I'm not really following... `SubscribePull` returns `c <-chan Descriptor, stop func()`, where `stop func()` is a function to deregister the returned channel from the subscriptions list of channels that should be notified

Reply:
If you stop a channel from the reader size and the writer tries to  write you get an error. A channel should be always closed by the writer.
Anyway, I still see the main reason to create PubSubChannel, if you don't see that this amount of code used in 3 places (already) must be extracted, maybe I have other threshold for code reuse.

Reply:
> If you stop a channel from the reader side and the writer tries to write, you get an error. A channel should be always closed by the writer.

@kortatu could you point to the code there this issue can happen. It is a very serious mistake if this is true, and we should fix it.

I cannot find the place were the channel is closed and what it is possible to write to it afterwards in SubscribePull or SubscribeToNeighbourhoodDepthChange.

> Anyway, I still see the main reason to create PubSubChannel, if you don't see that this amount of code used in 3 places (already) must be extracted, maybe I have other threshold for code reuse.

The method is the same in three places, but options and types are different. It can be extracted into a generalized subscription library, that would be a nice task with making sure that performance cost is minimal.

Reply:
In kademlia_load_balancing tests I found the problem on unsubscribing from the pubsub at the same time a new peer was added, but anyway, that is a Go design principle, only the writer close the channel.

The pubsub implementation has its own tests, that currently the other code doesn't.

But anyway I don't think I need to justify the refactor of some common functionality to an external file.

Reply:
@kortatu I do not understand the exact problem. Could you point to the code there you have found the problem with `SubscribeToNeighbourhoodDepthChange` and `SubscribePull`. It is very important that this issue is resolved as we rely on the that functionalities very much.

As far as I can see, every channel in these functions is owned by only one goroutine which removes the possibility of sent to a closed channel. Please, share our findings about the problems that you found there.

Also, as you say that `the subscriber can not stop the subscription from their side` is not true, as that functionality is provided by the returned unsubscribe function. Is it possible that I misunderstood your comment?

Reply:
> The pubsub implementation has its own tests, that currently the other code doesn't.
> 

Which tests are testing this pubsub implementation? I see no tests in gopubsub package.

And I think that we have tests for `SubscribePull` and `ToNeighbourhoodDepthChange`.

> But anyway I don't think I need to justify the refactor of some common functionality to an external file.

I am sorry if you feel offended. But, PR reviews because of any code change I think that it is ok to question any code or comment in PR review. Even, my questions are mainly about problems that you mentioned that you have found in other places.

I do not think that anybody has any objections on abstracting common functionality, I think that the abstraction implementation is in question here.

Reply:
I forgot to commit the test file. 
I found the problem with closing the subscription channel making the kademlia load balancing tests. As the addition of peers is done asynchronously, it could happen that the Load Balancer unsubscribe the moment a peer is added to Kademlia. Maybe is just a test thing, but is more correct to not close the channel from the reader side.

Reply:
I see what you are referring to for SubscribeToNeighbourhoodDepthChange, but since the channel is removed in the same function and there is syncOnce protecting the close, there should be no possibility for both send to the closed channel and close of the closed channel. This can be done with additional channel, not to close the returned channel from the client side but with more complexity. I found that this approach solved both problems in a simpler way since there already is Kademlia.lock.

In localstore SubscribePush a different approach is made and the returned channel is not closed on client side, but that subscription has different functionality then SubscribeToNeighbourhoodDepthChange.

Reply:
So it is working because the user of the API (the code subscribing) is doing things "kindly". I think an interface should not rely on that. As a generic publish/subscriber channel, I think that my implementation is safer.

Reply:
I have nothing to add. Thanks.